File: typescript/2023/src/day4/day4.ts
import { readAsStringArray } from "../utils/getInput";

const checkWinning = (input: string[]): number[] => {
  const prizeList: number[] = [];
  input
    .map((game) => game.split(":")[1].split("|"))
    .map((game) =>
      prizeList.push(
        checkCard(
          new Set(game[0].trim().split(/\s+/)),
          game[1].trim().split(/\s+/),
        ),
      ),
    );
  return prizeList;
};

const copyCards = (numTickets: number, prizeList: number[]) => {
  const cardCount = Array<number>(numTickets).fill(1);

  for (let i = 0; i < numTickets; i++) {
    const prize = prizeList[i];
    for (let j = 1; j <= prize; j++) {
      cardCount[i + j] = cardCount[i + j] + cardCount[i];
    }
  }

  return [...cardCount.values()].reduce((acc, val) => acc + val);
};

const checkCard = (winning: Set<string>, ticket: string[]) => {
  let points = 0;
  ticket.forEach((num) => {
    if (winning.has(num)) {
      points++;
    }
  });
  return points;
};

export const partOne = (filename: string): number => {
  const input = readAsStringArray(filename);
  const prizeList = checkWinning(input);
  return prizeList
    .map((prize) => {
      if (prize > 0) return Math.pow(2, prize - 1);
      else return 0;
    })
    .reduce((acc, val) => acc + val);
};

export const partTwo = (filename: string) => {
  const input = readAsStringArray(filename);
  const prizeList = checkWinning(input);
  return copyCards(input.length, prizeList);
};

File: golang/2023/day4/day4.go
package day4

import (
	"aoc/golang/utils"
	"fmt"
	// "math"
	"slices"
	"strings"
)

func checkWinning(input []string) (prizes []int) {
	for _, line := range input {
		game := strings.Split(line, ":")
		game = strings.Split(game[1], "|")
		winning, ticket := strings.Fields(game[0]), strings.Fields(game[1])

		wins := 0
		for _, num := range ticket {
			if slices.Contains(winning, num) {
				wins++
			}
		}
		prizes = append(prizes, wins)
	}

	return
}

func copyCards(prizes []int) (total int) {
	copies := make([]int, len(prizes))
	for i := range copies {
		copies[i] = 1
	}

	for i, prize := range prizes {
		for j := 1; j <= prize; j++ {
			copies[i+j] += copies[i]
		}
	}

	for _, c := range copies {
		total += c
	}
	return
}

func runPartOne(filename string) (points int) {
	fmt.Println("Part One")
	input := utils.ReadAsSlice[string](filename)
	for _, win := range checkWinning(input) {
		// points += int(math.Pow(2, float64(wins)-1))
		if win > 0 {
			points += 1 << (win - 1)
		}
	}
	return
}

func runPartTwo(filename string) int {
	fmt.Println("Part Two")
	input := utils.ReadAsSlice[string](filename)
	prizes := checkWinning(input)
	return copyCards(prizes)
}
